driver/src/orderbook/streamed/orderbook.rs

```diff
@@ -0,0 +1,85 @@
+use super::*;
+use crate::{
+    contracts::stablex_contract::batch_exchange,
+    models::{AccountState, Order},
+    orderbook::StableXOrderBookReading,
+};
+use anyhow::Result;
+use ethcontract::{contract::EventData, H256, U256};
+use state::{Batch, State};
+use std::collections::BTreeMap;
+
+// Ethereum events (logs) can be both created and removed. Removals happen if the chain reorganizes
+// and ends up not including block that was previously thought to be part of the chain.
+// However, the orderbook state (`State`) cannot remove events. To support this, we keep an ordered
+// list of all events based on which the state is built.
+
+/// The key by which events are sorted.
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct Key {
+    block_number: u64,
+    /// Is included to differentiate events from the same block number but different blocks which
+    /// can happen during reorgs.
+    block_hash: H256,
+    log_index: usize,
+}
+
+#[derive(Debug)]
+struct Value {
+    event: batch_exchange::Event,
+    /// The batch id is calculated based on the timestamp of the block.
+    batch_id: BatchId,
+}
+
+pub struct Orderbook {
```

Comment:
Would it make sense to give this struct a prefixed name such as `EventBasedOrderbook`?

Reply:
I don't think so because it is already in the streamed module.

Reply:
In general I also tend to prefer shorter names if the module clearly states what its for. Outside of this module we can use `streamed::Orderbook` so its clear that its a streamed version of the orderbook.